File: src/main/java/com/vc/medium/L1053.scala
package com.vc.medium

object L1053 {
  def prevPermOpt1(A: Array[Int]): Array[Int] = {
    val n = A.length
    var left = n - 2
    var right = n - 1
    while(left >= 0 && A(left) <= A(left + 1)) left -=1
    if(left < 0) return A
    while(A(left) <= A(right)) right -= 1
    while(A(right) == A(right - 1)) right -= 1
    val tmp = A(right)
    A(right) = A(left)
    A(left) = tmp
    A
  }
}

File: src/main/java/com/vc/medium/L1053.java
package com.vc.medium;

class L1053 {
    public int[] prevPermOpt1(int[] A) {
        if(A == null || A.length <= 1) return A;

        int n = A.length;
        int index1 = A.length - 2;
        while(index1 >= 0 && A[index1] <= A[index1 + 1]) index1--;

        if(index1 >= 0) {

            int index = index1 + 1;
            int index2 = index;
            while(index < n && A[index] < A[index1]) {
                if(A[index] > A[index2]) {
                    index2 = index;
                }
                index++;
            }

            int swap = A[index1];
            A[index1] = A[index2];
            A[index2] = swap;
        }

        return A;
    }
}
